feat(validations): add support to bind validation attributes

tolemac opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
Currently you can't bind `[minlength]` and `[maxlength]` directly. You can bind these as attributes using [attr.minlength] and [attr.maxlength] but the validatiors doesn't works.
You can bind `[pattern]` validator but the validator don't work correctly.
You can see: #10505 for details.

**What is the new behavior?**
You can bind `maxlength`, `minlength` and `pattern` using brackets or interpolation strings.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

**Other information**:

This change enables to bind the validations attributes `required`, `minlength`,
`maxlength` and `pattern`.

Closes: #10505, #7393

tolemac commented:
I have a problem updating the public api.

I'm getting one error running: `gulp public-api:update`
````
[12:39:46] Warning: gulp version mismatch:
[12:39:46] Global gulp is 3.9.1
[12:39:46] Local gulp is 3.9.0
[12:39:47] Using gulpfile E:\GitHub\tolemac\angular\gulpfile.js
[12:39:47] Starting 'build.sh'...
[12:47:01] Finished 'build.sh' after 7.23 min
[12:47:01] Starting 'public-api:update'...
E:\GitHub\tolemac\angular\node_modules\ts-api-guardian\build\lib\serializer.js:35
            throw new Error("Source file \"" + this.fileName + "\" not found");
            ^

Error: Source file "dist\packages-dist\core\index.d.ts" not found
    at ResolvedDeclarationEmitter.emit (E:\GitHub\tolemac\angular\node_modules\ts-api-guardian\build\lib\serializer.js:35:19)
    at publicApiInternal (E:\GitHub\tolemac\angular\node_modules\ts-api-guardian\build\lib\serializer.js:20:73)
    at Object.publicApi (E:\GitHub\tolemac\angular\node_modules\ts-api-guardian\build\lib\serializer.js:10:12)
    at Object.generateGoldenFile (E:\GitHub\tolemac\angular\node_modules\ts-api-guardian\build\lib\main.js:9:31)
    at Object.startCli (E:\GitHub\tolemac\angular\node_modules\ts-api-guardian\build\lib\cli.js:46:24)
    at Object.<anonymous> (E:\GitHub\tolemac\angular\node_modules\ts-api-guardian\bin\ts-api-guardian:3:29)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
[12:47:03] 'public-api:update' errored after 2.12 s
[12:47:03] Error: 1
    at formatError (C:\Users\jros\AppData\Roaming\npm\node_modules\gulp\bin\gulp.js:169:10)
    at Gulp.<anonymous> (C:\Users\jros\AppData\Roaming\npm\node_modules\gulp\bin\gulp.js:195:15)
    at emitOne (events.js:90:13)
    at Gulp.emit (events.js:182:7)
    at Gulp.Orchestrator._emitTaskDone (E:\GitHub\tolemac\angular\node_modules\orchestrator\index.js:264:8)
    at E:\GitHub\tolemac\angular\node_modules\orchestrator\index.js:275:23
    at finish (E:\GitHub\tolemac\angular\node_modules\orchestrator\lib\runTask.js:21:8)
    at cb (E:\GitHub\tolemac\angular\node_modules\orchestrator\lib\runTask.js:29:3)
    at ChildProcess.<anonymous> (E:\GitHub\tolemac\angular\gulpfile.js:93:33)
    at emitTwo (events.js:100:13)
````

It says `dist\packages-dist\core\index.d.ts` not found but the file exists.

Can someone help me to solve it?

vicb commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -97,11 +97,31 @@ export const MIN_LENGTH_VALIDATOR: any = {
+  @Input() minlength: string;
+
+  private _createValidator() {
+    this._validator = Validators.minLength(NumberWrapper.parseInt(this.minlength, 10));
```
pls remove NumberWrapper here

vicb commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -131,11 +151,31 @@ export const MAX_LENGTH_VALIDATOR: any = {
+  @Input() maxlength: string;
+
+  private _createValidator() {
+    this._validator = Validators.maxLength(NumberWrapper.parseInt(this.maxlength, 10));
```
same

tolemac commented:
@vicb NumberWrapper correction done!

tolemac committed:
feat(validations): add support to bind validation attributes

This change enables to bind the validations attributes `required`,
`minlength`, `maxlength` and `pattern`.

Closes: #10505, #7393

tolemac commented:
I have changed manually the public api in order to pass e2e build job in travis.

I have moved the problem with `gulp public-api:update` to [stackoverflow](http://stackoverflow.com/questions/38801573/building-angular2-error-runing-gulp-public-apiupdate)

tolemac commented:
@kara, what do you think about this PR?

kara commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -97,11 +97,31 @@ export const MIN_LENGTH_VALIDATOR: any = {
+    this._validator = Validators.minLength(parseInt(this.minlength, 10));
+  }
+
   constructor(@Attribute('minlength') minLength: string) {
```
Given that you have the @Input for "minlength", I'm curious why you kept the @Attribute injection around?  It doesn't seem necessary.  I think you should be able to remove it and lines 111-118 and instead rely on ngOnChanges.

tolemac replied:
True, true.
I'm agree I can remove the constructor from the three classes and rely on ngOnChanges.
I will do it.

kara commented on modules/@angular/forms/test/reactive_integration_spec.ts:
```diff
@@ -930,26 +932,94 @@ export function main() {
+               });
+             }));
+
+      it('should use sync validators using bindings',
```
This needs to be updated to use TestBed, as TCB has been removed from the repo.

kara commented on modules/@angular/forms/test/reactive_integration_spec.ts:
```diff
@@ -930,26 +932,94 @@ export function main() {
+                 fixture.debugElement.componentInstance.pattern = '.{3,}';
+                 fixture.detectChanges();
+
+                 var required = fixture.debugElement.query(By.css('[required]'));
```
const

kara commented on modules/@angular/forms/test/reactive_integration_spec.ts:
```diff
@@ -930,26 +932,94 @@ export function main() {
+                 dispatchEvent(required.nativeElement, 'input');
+                 dispatchEvent(minLength.nativeElement, 'input');
+                 dispatchEvent(maxLength.nativeElement, 'input');
+                 dispatchEvent(pattern.nativeElement, 'input');
```
Can you add a test that confirms that modifying the required, minLength, maxLength, and pattern properties after the initial check actually changes the validation state of the form?

kara commented:
@tolemac Can you also rebase this?

tolemac commented:
@kara, I have rebased it, I have removed the constructors in validator directives classes and I have added a test that confirms that when you change the validations properties the form state change.

I have errors on travis in build job `CI_MODE=e2e`.

````
1) 
MessageBroker should echo messages
  Message:
    Failed: unknown error: session deleted because of page crash
    from unknown error: cannot determine loading status
    from tab crashed
      (Session info: chrome=51.0.2704.0)
      (Driver info: chromedriver=2.20.353124 (035346203162d32c80f1dce587c8154a1efa0c3b),platform=Linux 3.13.0-40-generic x86_64)
  Stack:
    UnknownError: unknown error: session deleted because of page crash
    from unknown error: cannot determine loading status
    from tab crashed
      (Session info: chrome=51.0.2704.0)
      (Driver info: chromedriver=2.20.353124 (035346203162d32c80f1dce587c8154a1efa0c3b),platform=Linux 3.13.0-40-generic x86_64)
        at process._tickCallback (node.js:401:9)
    From: Task: <anonymous>
    From: Task: Waiting for element to be located By.cssSelector("app h1")
        at waitForBootstrap (../../../../../../modules/playground/e2e_test/web_workers/message_broker/message_broker_spec.ts:45:11)
        at Object.<anonymous> (../../../../../../modules/playground/e2e_test/web_workers/message_broker/message_broker_spec.ts:33:5)
    From: Task: Run it("should echo messages") in control flow
        at attemptAsync (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:1819:24)
        at QueueRunner.run (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:1774:9)
        at QueueRunner.execute (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:1762:10)
        at Spec.Env.queueRunnerFactory (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:627:35)
        at Spec.execute (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:353:10)
        at Object.fn (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:2360:37)
        at attemptAsync (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:1819:24)
        at QueueRunner.run (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:1774:9)
        at /home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:1801:16
    From asynchronous test: 
    Error
        at Suite.<anonymous> (../../../../../../modules/playground/e2e_test/web_workers/message_broker/message_broker_spec.ts:28:3)
        at addSpecsToSuite (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:743:25)
        at Env.describe (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:713:7)
        at jasmineInterface.describe (/home/travis/build/angular/angular/node_modules/jasmine-core/lib/jasmine-core/jasmine.js:3219:18)
        at Object.<anonymous> (../../../../../../modules/playground/e2e_test/web_workers/message_broker/message_broker_spec.ts:13:1)
````

I don't detect the cause ... can you help me to solve it?

kara commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -59,7 +57,12 @@ export const REQUIRED_VALIDATOR: any = {
 })
-export class RequiredValidator {
+export class RequiredValidator implements Validator {
+  @Input() required: string;
```
Can you make sure the "required" attribute is still present?  You may need a @HostBinding for that.

tolemac replied:
Ok, I will do the same for minlength, maxlength and pattern and test that attribute value is equal to binded property.

tolemac commented:
@kara, @HostBinding added and rebased ;)

tolemac commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -59,7 +57,12 @@ export const REQUIRED_VALIDATOR: any = {
+  @Input() @HostBinding('attr.required') required: string;
+
+  validate(c: AbstractControl): {[key: string]: any} {
+    return isPresent(this.required) ? Validators.required(c) : null;
```
hmn ... I have to test if its works when required=="false", ...
I think I have to test if `required` is present and not equal to "false" ...

tolemac replied:
... And `required` field should be boolean?

tolemac replied:
When required attribute is set to input like:

```` html
<input type="text" formControlName="login" required>
````

`required` value is set to empty string, however [W3C define it as `boolean attribute`](http://www.w3schools.com/tags/att_input_required.asp)

Here I miss something like [@BooleanFieldValue](https://github.com/angular/material2/blob/master/src/lib/core/annotations/field-value.ts) used in [Angular Material 2](https://github.com/angular/material2) for example [here](https://github.com/angular/material2/blob/master/src/lib/input/input.ts#L119), this is very useful to automatically convert input properties to boolean values.

I don't know if that exists in angular 2 source code therefore I will do it manually in required validator.

tolemac commented:
I have change `required` type to boolean as I self discussed [here](https://github.com/angular/angular/pull/10518#discussion-diff-76384702L62)

Now, when `required` is "false", it isn't call to required validator.
I have add test that confirm it.

tolemac commented:
I have added tests to confirm that validation attributes aren't assigned to host element when bound properties are falsy.

kara commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -6,9 +6,9 @@
  */
 
-import {Attribute, Directive, forwardRef} from '@angular/core';
+import {Attribute, Directive, HostBinding, Input, OnChanges, SimpleChanges, forwardRef} from '@angular/core';
```
Remove Attribute and HostBinding, as they're no longer used

kara commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -57,9 +55,20 @@ export const REQUIRED_VALIDATOR: any = {
+  @Input()
+  get required(): boolean { return this._required; }
+
+  set required(value: boolean) { this._required = isPresent(value) && `${value}` !== 'false'; }
```
It seems a bit weird to me that some falsy values like `null` and `undefined` evaluate to required: false, while others like `0` and `''` evaluate to true.  It might make more sense to use the same pattern as `disabled` in `ngModel`: `value != null && value != false`

kara commented on modules/@angular/forms/test/reactive_integration_spec.ts:
```diff
@@ -933,37 +933,175 @@ export function main() {
+        dispatchEvent(required.nativeElement, 'input');
+        dispatchEvent(minLength.nativeElement, 'input');
+        dispatchEvent(maxLength.nativeElement, 'input');
+        dispatchEvent(pattern.nativeElement, 'input');
```
You shouldn't have to type into the form to update the validation state.  If it's necessary to emit these events to make the test work, something isn't hooked up fully. 

It seems to me that `updateValueAndValidity()` isn't being called on the control in question, so changing the value of `required`, etc won't be enough to change the validation state of the form.  You'll need to add a callback or something similar to ensure the control's validity state is updated as soon as these values change.    You'll also need to keep in mind that for reactive controls, the control bound to the directive can change, so any callback will need to be cleared on rebinding.

kara commented on modules/@angular/forms/test/reactive_integration_spec.ts:
```diff
@@ -933,37 +933,175 @@ export function main() {
+        expect(form.valid).toEqual(true);
+      });
+
+      it('should use sync validators using bindings', () => {
```
Can you add these same tests to template_integration_spec?  It's more common and encouraged to use validator directives with template-driven (with reactive, you should really be editing the validator list programmatically).

kara commented:
@tolemac You've done a great job on this PR.  I'm realizing that this feature will require larger changes to how validator directives work, though. To ensure we can get this feature in for RC.6, I can take it from here. I've pulled down your work and will build on top of it (don't worry, I will keep your commit to give you credit :) ).

kara commented:
Included your commit in my PR here: https://github.com/angular/angular/pull/11116.
